README.md
# Snakes AI Genetic Algorithm Neural Network

## To setup in Python 3 environment run:
`pip3 install requirements.txt`

## To run the best trained neural network run:

`python3 trainSimpNoFitness.py`

## To run the snake game in manual mode run:

`python3 main.py`


### Note:
I created this project 4 years ago in 2019 as I write this, so I don't remember how to do all of the traning functions. Feel free to figure it out

History.txt
Dec 7:
- Trained Large network of whole board input 10,000 times, didn't get any better.
- Learned of NEAT algorithm, where the topology of the network changes too. super interesting.
Dec 6 2018:
- finshed training 3 different sizes over 5000 generations. they all still stopped around 20, and the snakes still run into
	themselves? How is this possible?
- Trying different fitness functions: v1: normal, v2: award for apple, award for moving toward, v3: only award for apple
- (dec 7) Tested with same random seed to eliminate differences. v2 maxed out lower. v1 and v3 were similar, just v3 was a
	little slower to getting up to that 20.5 average high. I think v3 has more potential, it didn't randomly kill itself
	like it has with other version
Dec 5 2018:
- finished training 24000 generations over night! took maybe 12 hours. After 3000 generations, it didn't get any better!
	It maxed its score around 25, averaging 20. My goal is average of 50. I think I need to make the network bigger.
	Also might try if something is right in front of it be 0 distance, not 1
- Nate Ludlow gave the idea of training 3-4 different sized networks at the same time to see which is better, just 
	for 5000 generations each. can multiprocess.
Dec 4 2018:
- Created load function for AIv0.2
- Finshed input structuring for simplified network
- Trained for 500 generations with simplified network and got to a length of 15!!!!!
Dec 1 2018:
- Changed AIv0.2 to account for subclasses when crossing over. returns an instance of the subclass if there is one now.
- Started working on the evolve function, with multiple populations that migrate, but train independently
- train() takes a long time. 47 sec for 100 gen with 8 crossovers, 40 sec w/o crossover
Nov 29-30:
- Brutal data alayzing. Lots of running through it and printing. Fitness somehow goes to 0 for all on second generation,
	and after, always, even if length is 2.
- Somehow the snake pos if often off the board???
- I think I fixed the zero error, through reordering the play_game function. 
- Still issue of player way off of the board- (34,8)
- FOUND ISSUE OF GOING OFF BOARD!! in the train function, it would play a game 3 times, but it wouldn't re-intialize
	the values of pos, direction etc. They carried over into the next game
Nov 28:
- Fixed play_game. Created a timeit function for play_game. Can do 1000 games in a about 4.3 seconds.
- Finished GUI, as far as I know. Doesn't come up front for windows though. Too lazy to fix
Nov 27:
- made progress in play_game, just need to adjust the output it returns a list of the output neurons, 
	instead of which one was the max
Nov 17:
	-tested memory size of 50 neural nets for the size 192x16x16x4. If my calculations
		are right, it should only take a few MB!!

Nov 8:
 - Completely rewrote the AI class as Network - A lot cleaner and more effective
Sep 22:
 - Switched keyup to keydown, makes a lot more responsive

Sep 21:
 - Made arena resizable.


Before:
 - created backlog of buttons pressed for double moves

file.py
from pickle import dump, load
from os import walk, getcwd

#Tested Nov 9 2018, faster then numpy.save for arrays and integers!!!!
__author__ = "Anton Alley"
__version__ = "0.0.0"
__copyright__ = "Copyright 2018 Anton Alley all rights reserved"

class OverwriteError(Exception):
    def __init__(self,value):
        self.value = value

    def __str__(self):
        return repr(self.value)

def save(filename, data, warning=False, avoidOverwrite=True):
    if warning:
        input('Warning: will overwite entire file. Press enter to continue')

    if avoidOverwrite:
        filenames = []
        for (_,_,f) in walk(getcwd()):
            filenames.extend(f)
        if f'{filename}.sav' in  filenames:
            raise OverwriteError("Attempted to overwrite a file")
            return None
    
    
    with open(filename+'.sav', 'wb') as f:
        dump(data, f)


def f_open(filename):
    with open(filename+'.sav', 'rb') as f:
        return load(f)


main.py
#!/usr/bin/env python
# -*- coding: <ASCII> -*-
# main.py


#################################
# ABOUT:
#
# The main file for snake game
# Run This file to manualy play the game
#
#################################

__author__ = "Anton Alley"
__copyright__ = "None"
__credits__ = ["Anton Alley"] 
__license__ = "None"
__version__ = "0.0.0"
__maintainer__ = "Anton Alley"
__email__ = "Anton.Alley@gmail.com"

# Builtin Imports
from random import randint
import time
import sys, os

# Third Party Imports
import pygame
from pygame.locals import *

# My Imports


# USE GRAPHICS:
##from platform import platform
##from os import environ
##
##if 'Windows' in platform():
##    environ['SDL_VIDEODRIVER'] = 'directx'


######### CONSTANTS:
FPS =           8 # the speed of the game
ARENA_SIZE =    (12,16) # DEFAULT FOR AI DON'T CHANGE!!
SQUARE_SIZE =   45
BOX_THICKNESS = int(SQUARE_SIZE / 10)

EMPTY = 0
SNAKE = 1
APPLE = 2

LIGHT_GRAY = (100,100,100)
PALE_RED   = (200, 40, 40)
WHITE      = (255,255,255)
###############

os.environ["SDL_VIDEO_CENTERED"] = '1' # Centers Screen


class player:
    def __init__(self):
        self.pos = (1, int(ARENA_SIZE[1] / 2))
        self.length = 1
        self.direction = (1,0)

    def getDir(self, direction):
        if direction == 'up':
            return (-1,0)
        elif direction == 'down':
            return (1,0)
        elif direction == 'left':
            return (0,-1)
        elif direction == 'right':
            return (0,1)
        elif direction == 'turnRight':
            tempDir = self.direction
            if tempDir[0] != 0:
                tempDir = (tempDir[0] * -1, tempDir[1])
            return (tempDir[1],tempDir[0])
        elif direction == 'turnLeft':
            tempDir = self.direction
            if tempDir[1] != 0:
                tempDir = (tempDir[0], tempDir[1] * -1)
            return (tempDir[1],tempDir[0])
        elif direction == 'continueStraight':
            return self.direction
        else:
            raise Exception(f"False Direction Given, no: {direction}")
        
    def move(self, direction):
        self.direction = self.getDir(direction)
            
    def update(self):
        posUD = self.direction[0] + self.pos[0]
        posLR = self.direction[1] + self.pos[1]
        self.pos = (posUD,posLR)



class gameMap:
    def __init__(self, snake):
        player.__init__(snake) # Re-intialize position, length etc. In case it changed

        self.grid = [[0 for i in range(ARENA_SIZE[1])] for i in range(ARENA_SIZE[0])]
        # 0 is empty, 1 is snake, 2 is apple
        self.apple = snake.pos
        while self.apple == snake.pos: # So apple doesn't overlap initially with snake
            self.apple = (randint(0,ARENA_SIZE[0]-1),randint(0,ARENA_SIZE[1]-1))
        self.grid[self.apple[0]][self.apple[1]] = APPLE
        
        self.ticker = []
        self.frame = 0
        self.update(snake)

    def __str__(self):
        toprint = ""
        for i in self.grid:
            for a in i:
                toprint += str(a)+" "
            toprint += "\n"
        print(toprint)
        return ""

    def update(self,snake):
        snake.update()
        # New begginning of snake, or DEATH:
        try:
            if snake.pos[0] >= 0 and snake.pos[0] < ARENA_SIZE[0] and snake.pos[1] >= 0 and snake.pos[1] < ARENA_SIZE[1] and self.grid[snake.pos[0]][snake.pos[1]] != SNAKE:
                    self.grid[snake.pos[0]][snake.pos[1]] = SNAKE
            else:
                return "ENDGAME", snake.length
        except:
            return "ENDGAME", snake.length
        

        # UPDATE SNAKE:
        self.ticker.append([snake.pos,self.frame])
        toDeleteIndicies = []
        for index, box in enumerate(self.ticker): # delete the end of the snake
            if self.frame - box[1] >= snake.length:
                self.grid[box[0][0]][box[0][1]] = EMPTY
                toDeleteIndicies.append(index)  #deletes the old positions that it has moved out of
                
        for index in toDeleteIndicies:
            del self.ticker[index]
            

        # APPLE SPAWN:
        if snake.pos == self.apple:
            self.apple = (randint(0,ARENA_SIZE[0]-1),randint(0,ARENA_SIZE[1]-1))
            snake.length += 1
            # So the apple can't spawn where the snake is:
            while self.grid[self.apple[0]][self.apple[1]] == 1:
                self.apple = (randint(0,ARENA_SIZE[0]-1),randint(0,ARENA_SIZE[1]-1))
            self.grid[self.apple[0]][self.apple[1]] = 2
                
                
        self.frame += 1
        return None


def test():
    s = player()
    g = gameMap(s)
    str(g)
    time.sleep(3)
    s.move("right")
    g.update(s)
    str(g)
    time.sleep(3)
    s.move('down')
    g.update(s)
    str(g)
    


def main():
    ######## Game Setup:
    s = player()
    g = gameMap(s)
    WIDTH, HEIGHT = ((SQUARE_SIZE * ARENA_SIZE[1]) , (SQUARE_SIZE * ARENA_SIZE[0]))
    ########
    
    DISPLAY = pygame.display.set_mode((WIDTH,HEIGHT),0,32)
    pygame.display.set_caption(" ~SNAKE~")
    pygame.init()
    nextChangeDirection = []
    gameClock = pygame.time.Clock()
    pause = False
    while True:
        DISPLAY.fill((0,0,0))

        
        for event in pygame.event.get():
            if event.type == KEYDOWN:
                if event.key == K_DOWN:
                    nextChangeDirection.append("down")
                    #s.move("down")
                if event.key == K_UP:
                    nextChangeDirection.append("up")
                    #s.move("up")
                if event.key == K_RIGHT:
                    nextChangeDirection.append("right")
                    #s.move("right")
                if event.key == K_LEFT:
                    nextChangeDirection.append("left")
                    #s.move("left")
                if event.key == K_p:
                    pause = not pause
                
            elif event.type == QUIT:
                print(f"The snake had a final length of: {s.length}")
                pygame.quit()
                sys.exit()


       
        ##################### The game:
        if not pause:
            if nextChangeDirection != []:
                s.move(nextChangeDirection[0])
                del nextChangeDirection[0]
            result = g.update(s)
            if result != None:
                print(f"The snake had a final length of: {result[1]}")
                if result[1] < 15:
                    print("Nice going stupid!")
                else:
                    print("Great Job!")
                pygame.quit()
                sys.exit()
        for c, C in enumerate(g.grid):
            for r, R in enumerate(C):
                if R == 1:  # Snake
                    rect = ((r * SQUARE_SIZE, c * SQUARE_SIZE), (SQUARE_SIZE, SQUARE_SIZE))
                    pygame.draw.rect(DISPLAY, WHITE, rect)
                elif R == 2: # Apple
                    rect = ((r * SQUARE_SIZE, c * SQUARE_SIZE), (SQUARE_SIZE, SQUARE_SIZE))
                    pygame.draw.rect(DISPLAY, PALE_RED, rect)

        #########################
        

        # The grid:
        for R in range(ARENA_SIZE[1]):
            for D in range(ARENA_SIZE[0]):
                rect = ((R * SQUARE_SIZE, D * SQUARE_SIZE), (SQUARE_SIZE, SQUARE_SIZE))
                pygame.draw.rect(DISPLAY, LIGHT_GRAY, rect, BOX_THICKNESS)

        pygame.display.flip()
        gameClock.tick(FPS)

    

                
    return None


if __name__ == "__main__":
    main()


AI_v0_2.py
#!/usr/bin/env python
# -*- coding: <ASCII> -*-
# AI_v0_2.py


#################################
# ABOUT:
# - Just got a fresh start. Trying to make code cleaner
# - Using numpy arrays for optimization. 
# - Has custom shape to neural network
# - Mutation is based on Bell curve distribution
#
#################################

__author__ = "Anton Alley"
__copyright__ = "None"
__credits__ = ["Anton Alley"] 
__license__ = "None"
__version__ = "0.2.0"
__maintainer__ = "Anton Alley"
__email__ = "Anton.Alley@gmail.com"

# Builtin Imports
import math
import random

# Third Party Imports
import numpy as np

# My Imports
import file as fileHandler

#Global Variables:
MUTATION_RATE = 0.07
MUTATION_RANGE = 0.25
NETWORK_STORAGE_LOCATION = "networks/Dec112018/"

def ReLU(x):
    return max([0, x])

class NetworkError(Exception):
    def __init__(self, value):
        self.value = value
        
    def __str__(self):
        return repr(self.value)


class Network:
    def __init__(self, networkShape=None, file=None):
        '''Uploads File if proper format, else use networkShape as a list of the numer of neurons in each layer, including output and input layers'''
        if file == None and networkShape != None:   
            if type(networkShape) != list:
                raise NetworkError(f"Wrong network shape type: {type(networkShape)}")
            else:
                self.newNetwork(networkShape)
        elif file != None:
            self.networkShape, self.weights, self.biases = fileHandler.f_open(file)
            self.file = file
        else:
            return None

        numBiases = sum(self.networkShape[1:])
        numWeights = sum(networkShape[i - 1] * networkShape[i] for i in range(len(networkShape[1:])))
        self.numDials = numBiases + numWeights

        
    def newNetwork(self, networkShape):
        if len(networkShape) < 2:
            raise NetworkError(f"Must have at least 2 layers. You had: {len(networkShape)}")
        if all([1 if a > 0 else 0 for a in networkShape]) == False:
            raise AIError("Number of neurons must be greater then 0")
        
        self.networkShape = networkShape
        
        self.biases = []
        self.weights = []
        for numNeurons, incomingNeurons in zip(networkShape[1:], networkShape):
            self.weights.append(np.random.uniform(-0.5,0.5,size = (numNeurons, incomingNeurons)))
            self.biases.append(np.random.uniform(-5,6,size = (numNeurons)))

    def openData(self, networkShape, weights, biases):
        self.networkShape = networkShape
        self.weights = weights
        self.biases = biases
                         
    def feedForward(self, inputVector):
        if len(inputVector) != self.networkShape[0]:
            raise NetworkError(f"Wrong size for inputVector: {len(inputVector)}, must be a length of {self.networkShape[0]}")
        for data in inputVector:
            if type(data) != float and type(data) != int:
                raise NetworkError(f"Wrong data type in inputVector: {type(data)}; must by int or float")
        
        forwardData = inputVector
        for b, w in zip(self.biases, self.weights):
            forwardData = [ReLU(i) for i in (np.dot(w, forwardData) + b)]
        return forwardData

    def save(self):
        while 1: # to avoid repeats
            try:
                r_ID = str(random.randint(1000000,9999999))
                fileHandler.save(f"{NETWORK_STORAGE_LOCATION}{r_ID}",[self.networkShape, self.weights, self.biases])
            except fileHandler.OverwriteError:
                continue
            else:
                return r_ID
    def load(self, ID):
        inData = fileHandler.f_open(NETWORK_STORAGE_LOCATION + ID)
        self.networkShape, self.weights, self.biases = inData

    def mutate_DEP(self):
        maxMutations = max(1, round(self.numDials * MUTATION_RATE))
        numMutations = random.randint(1, maxMutations)

        for _ in range(numMutations):
            layerToMute = random.randint(0,len(self.networkShape) - 2)
            neuronToMute = random.randint(0, self.networkShape[layerToMute+1] - 1)
            # -1 means bias, else is the index
            dataToMute = random.randint(-1, self.networkShape[layerToMute] - 1)

            muteAmount = random.uniform(-MUTATION_RATE, MUTATION_RATE)
            if dataToMute != -1:
                self.weights[layerToMute][neuronToMute][dataToMute] += muteAmount
            else:
                self.biases[layerToMute][neuronToMute] += muteAmount * 8 # Biases should swing more then the weightes, hense *8


    def mutate(self):
        for Li, layer in enumerate(self.weights):
            for Ni, neuron in enumerate(layer):
                for Wi, weight in enumerate(neuron):
                    if random.uniform(0,1) < MUTATION_RATE:
                        self.weights[Li][Ni][Wi] += np.random.normal(0, MUTATION_RANGE/2)

        for Li, layer in enumerate(self.biases):
            for Bi, bias in enumerate(layer):
                if random.uniform(0,1) < MUTATION_RATE:
                    self.biases[Li][Bi] += np.random.normal(0,MUTATION_RANGE)# Biases should swing more then the weightes so not /2

        
    @staticmethod
    def crossover(net1, net2, version = "arithmetic"):
        '''Does a cross over with the data given. version can = 'uniform' OR 'arithmetic' OR 'random' '''
        if net1.networkShape != net2.networkShape:
            raise NetworkError(f"Cannot crossover 2 different sized Networks\nAI1: {AI1.networkShape}\nAI2: {AI2.networkShape}")
        if version not in ["uniform", "arithmetic", "random"]:
            raise AIError(f"The {version} version of crossover is supported")
        if version == "random":
            version = r.choice(["uniform", "arithmetic"])

        childBiases = []
        childWeights = []
        if version == "arithmetic":
            net1_weight = random.uniform(0.1, 0.9)
            net2_weight = 1 - net1_weight
            childWeights = [w1+w2 for w1, w2 in zip(
                (w * net1_weight for w in net1.weights),
                (w * net2_weight for w in net2.weights)
                )]
            childBiases = [b1 + b2 for b1, b2 in zip(
                (b * net1_weight for b in net1.biases),
                (b * net2_weight for b in net2.biases)
                )]
        elif version == "uniform":
            raise NetworkError("I am too lazy to program this do it later!!!")

        newNetwork = type(net1)(None) #Network(None), but returns subclass type if in use
        newNetwork.openData(net1.networkShape, childWeights, childBiases)

        return newNetwork
    


dep/speedTest.py
#!/usr/bin/env python
# -*- coding: <ASCII> -*-
# AI_v0_2.py


#################################
# ABOUT:
# - Just got a fresh start. Trying to make code cleaner
# - Using numpy arrays for optimization. 
# - Has custom shape to neural network
# - Mutation is based on Bell curve distribution
#
#################################

__author__ = "Anton Alley"
__copyright__ = "None"
__credits__ = ["Anton Alley"] 
__license__ = "None"
__version__ = "0.2.0"
__maintainer__ = "Anton Alley"
__email__ = "Anton.Alley@gmail.com"

# Builtin Imports
import math
import random

# Third Party Imports
import numpy as np

# My Imports
import file as fileHandler

#Global Variables:
MUTATION_RATE = 0.07
MUTATION_RANGE = 0.25
NETWORK_STORAGE_LOCATION = "networks/"

def ReLU(x):
    return max([0, x])

class NetworkError(Exception):
    def __init__(self, value):
        self.value = value
        
    def __str__(self):
        return repr(self.value)


class Network:
    def __init__(self, networkShape=None, file=None):
        '''Uploads File if proper format, else use networkShape as a list of the numer of neurons in each layer, including output and input layers'''
        if file == None and networkShape != None:   
            if type(networkShape) != list:
                raise NetworkError(f"Wrong network shape type: {type(networkShape)}")
            else:
                self.newNetwork(networkShape)
        elif file != None:
            self.networkShape, self.weights, self.biases = fileHandler.f_open(file)
        else:
            return None

        numBiases = sum(self.networkShape[1:])
        numWeights = sum(networkShape[i - 1] * networkShape[i] for i in range(len(networkShape[1:])))
        self.numDials = numBiases + numWeights

        
    def newNetwork(self, networkShape):
        if len(networkShape) < 2:
            raise NetworkError(f"Must have at least 2 layers. You had: {len(networkShape)}")
        if all([1 if a > 0 else 0 for a in networkShape]) == False:
            raise AIError("Number of neurons must be greater then 0")
        
        self.networkShape = networkShape
        
        self.biases = []
        self.weights = []
        for numNeurons, incomingNeurons in zip(networkShape[1:], networkShape):
            self.weights.append(np.random.uniform(-0.5,0.5,size = (numNeurons, incomingNeurons)))
            self.biases.append(np.random.uniform(-5,6,size = (numNeurons)))

    def openData(self, networkShape, weights, biases):
        self.networkShape = networkShape
        self.weights = weights
        self.biases = biases
                         
    def feedForward(self, inputVector):
        if len(inputVector) != self.networkShape[0]:
            raise NetworkError(f"Wrong size for inputVector: {len(inputVector)}, must be a length of {self.networkShape[0]}")
        for data in inputVector:
            if type(data) != float and type(data) != int:
                raise NetworkError(f"Wrong data type in inputVector: {type(data)}; must by int or float")
        
        forwardData = inputVector
        for b, w in zip(self.biases, self.weights):
            forwardData = [ReLU(i) for i in (np.dot(w, forwardData) + b)]
        return forwardData

    def save(self):
        while 1: # to avoid repeats
            try:
                r_ID = str(random.randint(1000000,9999999))
                fileHandler.save(f"{NETWORK_STORAGE_LOCATION}{r_ID}",[self.networkShape, self.weights, self.biases])
            except fileHandler.OverwriteError:
                continue
            else:
                return r_ID
    def load(self, ID):
        inData = fileHandler.f_open(NETWORK_STORAGE_LOCATION + ID)
        self.networkShape, self.weights, self.biases = inData

    def mutate_DEP(self):
        maxMutations = max(1, round(self.numDials * MUTATION_RATE))
        numMutations = random.randint(1, maxMutations)

        for _ in range(numMutations):
            layerToMute = random.randint(0,len(self.networkShape) - 2)
            neuronToMute = random.randint(0, self.networkShape[layerToMute+1] - 1)
            # -1 means bias, else is the index
            dataToMute = random.randint(-1, self.networkShape[layerToMute] - 1)

            muteAmount = random.uniform(-MUTATION_RATE, MUTATION_RATE)
            if dataToMute != -1:
                self.weights[layerToMute][neuronToMute][dataToMute] += muteAmount
            else:
                self.biases[layerToMute][neuronToMute] += muteAmount * 8 # Biases should swing more then the weightes, hense *8


    def mutate(self):
        for Li, layer in enumerate(self.weights):
            for Ni, neuron in enumerate(layer):
                for Wi, weight in enumerate(neuron):
                    if random.uniform(0,1) < MUTATION_RATE:
                        self.weights[Li][Ni][Wi] += np.random.normal(0, MUTATION_RANGE/2)

        for Li, layer in enumerate(self.biases):
            for Bi, bias in enumerate(layer):
                if random.uniform(0,1) < MUTATION_RATE:
                    self.biases[Li][Bi] += np.random.normal(0,MUTATION_RANGE)# Biases should swing more then the weightes so not /2

        
    @staticmethod
    def crossover(net1, net2, version = "arithmetic"):
        '''Does a cross over with the data given. version can = 'uniform' OR 'arithmetic' OR 'random' '''
        if net1.networkShape != net2.networkShape:
            raise NetworkError(f"Cannot crossover 2 different sized Networks\nAI1: {AI1.networkShape}\nAI2: {AI2.networkShape}")
        if version not in ["uniform", "arithmetic", "random"]:
            raise AIError(f"The {version} version of crossover is supported")
        if version == "random":
            version = r.choice(["uniform", "arithmetic"])

        childBiases = []
        childWeights = []
        if version == "arithmetic":
            net1_weight = random.uniform(0.1, 0.9)
            net2_weight = 1 - net1_weight
            childWeights = [w1+w2 for w1, w2 in zip(
                (w * net1_weight for w in net1.weights),
                (w * net2_weight for w in net2.weights)
                )]
            childBiases = [b1 + b2 for b1, b2 in zip(
                (b * net1_weight for b in net1.biases),
                (b * net2_weight for b in net2.biases)
                )]
        elif version == "uniform":
            raise NetworkError("I am too lazy to program this do it later!!!")

        newNetwork = type(net1)(None) #Network(None), but returns subclass type if in use
        newNetwork.openData(net1.networkShape, childWeights, childBiases)

        return newNetwork



def speed(num):
    import time
    x = Network([4,24,24,24,24,3])
    t = time.clock()
    for i in range(num):
        x.mutate()

    return time.clock() - t


    
    


trainOG.py
#!/usr/bin/env python
# -*- coding: <ASCII> -*-
# train.py

#####CURRENTLY WORKING ON:
#
# 


#################################
# ABOUT:
#
#
#
# This module is to train the AI for the snakes game
#
#
#
#################################

__author__ = "Anton Alley"
__copyright__ = "None"
__credits__ = ["Anton Alley"] 
__license__ = "None"
__version__ = "0.0.0"
__maintainer__ = "Anton Alley"
__email__ = "Anton.Alley@gmail.com"

# Builtin Imports
import random
import time
import sys, os
import math
from copy import deepcopy

# Third Party Imports
import pygame
from pygame.locals import *

# My Imports
import main as SG
from AI_v0_2 import Network as BACKGROUND_AI


######### CONSTANTS:
SG.ARENA_SIZE = (12, 12)
BOARD_SIZE = 144
ADD_VARIABLES = 7 #+ Head pos(2), Apple pos(2), Length(1), currDir(2)
INPUT_DATA_SIZE = BOARD_SIZE + ADD_VARIABLES
###############

class SnakeAi(BACKGROUND_AI, SG.player):
    def __init__(self, networkShape=[INPUT_DATA_SIZE, 30, 36, 36, 30, 27, 21, 15, 9, 3], file = None):
        BACKGROUND_AI.__init__(self, networkShape=networkShape, file=file)
        SG.player.__init__(self)

    def GetMove(self, inputData):
        vect_board = []
        for col, r in enumerate(inputData[:-1]):
            for data in r:
                vect_board.append(data)
        for d in inputData[-1]:
            vect_board.append(d)
                
        output = self.feedForward(vect_board)
        resultingMove = output.index(max(output))
        if resultingMove == 0:
            return "continueStraight"
        elif resultingMove == 1:
            return "turnRight"
        elif resultingMove == 2:
            return "turnLeft"

        else:
            raise Exception(f"IMPROPER OUTPUT FROM the Neural Network: {output}")
        

def distance(p1, p2):
    return math.sqrt((p2[0]-p1[0])**2 + (p2[1]-p1[1])**2)
    
def play_game(snakePlayer):
    game        = SG.gameMap(snakePlayer)

    currentFrameNumber = 0
    frameOfLastLengthChange = 0
    previousLength = 1
    while True:
        snakePlayer.move(snakePlayer.GetMove(game.grid + [[*snakePlayer.pos,*snakePlayer.direction,
                                                     *game.apple, snakePlayer.length]]))
        result = game.update(snakePlayer)

        if snakePlayer.length > previousLength:
            frameOfLastLengthChange == currentFrameNumber
            previousLength = snakePlayer.length


        ######Stop if stopped:
        if currentFrameNumber - frameOfLastLengthChange > BOARD_SIZE:
            return snakePlayer.length
            
        if result != None:
            finalLength = result[1]

            return finalLength


            
        currentFrameNumber += 1
        

    
def train(numGens, batchSize, numBatchWinners,
          batch = None, extraLosers=0, numCrossovers=0):

    if batch is None:
        batch = []
    else:
        if len(batch) > batchSize:
            raise Exception("Batch given too big")
        
    NUM_TEST_GAMES = 1

    LENGTH_PROGRESSION = []
    HIGH_LENGTH = []
    LOW_LENGTH = []

    if batch == []:
        for i in range(batchSize):
            batch.append(SnakeAi())

    for genNum in range(numGens):
        if genNum % 10 == 0:
            print('Progress:', genNum, '/',numGens)
        lengths = []
        for Ni, network in enumerate(batch):
            for i in range(NUM_TEST_GAMES):
                length = play_game(network)
                if i == 0:
                    lengths.append(length)
                lengths[Ni] = (length + lengths[Ni]) / 2

                
        LENGTH_PROGRESSION.append(sum(lengths) / len(lengths))
        HIGH_LENGTH.append(max(lengths))
        LOW_LENGTH.append(min(lengths))
        
        winners = []
        for i in range(numBatchWinners):
            currHighIndex = lengths.index(max(lengths))
            winners.append(batch.pop(currHighIndex))
            lengths.pop(currHighIndex)
        
        
        
          
        if genNum < numGens - 1:
            for loser in batch[:extraLosers]:
                winners.append(loser)

            batch = deepcopy(winners)
        
            for child in range(numCrossovers):
                batch.append(
                    SnakeAi.crossover(random.choice(winners), random.choice(winners))
                    )
            
            for i in range(batchSize - numBatchWinners - numCrossovers):
                netToMutate = deepcopy(random.choice(batch))
                netToMutate.mutate()
                batch.append(netToMutate)
                

    return winners,[LENGTH_PROGRESSION, HIGH_LENGTH, LOW_LENGTH]
            


def evolve(numPopulations, epochs):
    NUM_GEN_PER_EPOCH = 100
    NUM_BATCH_WINNERS = 10
    populations = [None for i in range(numPopulations)]
    popSettings = [{"xtraLosers":random.randint(0, 3), "numCrossovers":random.randint(3, 15), "batchSize":random.randint(40,65)} for i in range(numPopulations)]
    popLengthHistory = [[] for i in range(numPopulations)]
    
    for e in range(epochs):
        for pI in range(numPopulations):
            rBatch, lengths = train(NUM_GEN_PER_EPOC, popSettings[pI]["batchSize"], NUM_BATCH_WINNERS,
                                                batch = populations(pI), extraLosers = popSettings[pI]["xtraLosers"], numCrossovers = popSettings[pI]["numCrossovers"])
            populations.append(rBatch)

            for length in lengths[1]:
                popLengthHistory[pI].append(length)

        ##### Migrate:
        movers = []
        for population in populations:
            # 1 migrater per population, which has NUM_BATCH_WINNERS each
            movers.append(population.pop(random.randint(0, len(population)-1)))
        random.shuffle(movers)
        for i, snake in enumerate(movers):
            populations[i].append(snake)
        del movers
        ######
    return populations, popLengthHistory
        

def gui(AI):

    FPS =           10 # the speed of the game
    ARENA_SIZE =    (12,16) # DEFAULT FOR AI DON'T CHANGE!!
    SQUARE_SIZE =   45
    BOX_THICKNESS = int(SQUARE_SIZE / 10)

    LIGHT_GRAY = (100,100,100)
    PALE_RED   = (200, 40, 40)
    WHITE      = (255,255,255)
    BLACK      = (  0,  0,  0)
    os.environ["SDL_VIDEO_CENTERED"] = '1' # Centers Screen

    ######## Game Setup:
    s = AI
    game = SG.gameMap(s)
    WIDTH, HEIGHT = ((SQUARE_SIZE * ARENA_SIZE[1]) , (SQUARE_SIZE * ARENA_SIZE[0]))
    ########

    DISPLAY = pygame.display.set_mode((WIDTH,HEIGHT),0,32)
    pygame.display.set_caption(" ~SNAKE~")
    pygame.init()
    nextChangeDirection = []
    gameClock = pygame.time.Clock()
    while True:
        DISPLAY.fill(BLACK)

        for event in pygame.event.get():
            if event.type == QUIT:
                print(f"The snake had a final length of: {s.length}")
                pygame.quit()
                sys.exit()



        ###THE Game:
        s.move(s.GetMove(game.grid + [[*s.pos,*s.direction,*game.apple, s.length]]))
        result = game.update(s)
        if result != None:
            print(f"The snake had a final length of: {result[1]}")
            pygame.quit()
            return None
            #sys.exit()

        for c, C in enumerate(game.grid):
            for r, R in enumerate(C):
                if R == 1:  # Snake
                    rect = ((r * SQUARE_SIZE, c * SQUARE_SIZE), (SQUARE_SIZE, SQUARE_SIZE))
                    pygame.draw.rect(DISPLAY, WHITE, rect)
                elif R == 2: # Apple
                    rect = ((r * SQUARE_SIZE, c * SQUARE_SIZE), (SQUARE_SIZE, SQUARE_SIZE))
                    pygame.draw.rect(DISPLAY, PALE_RED, rect) 

        ####



        # The Grid:
        for R in range(ARENA_SIZE[1]):
            for D in range(ARENA_SIZE[0]):
                rect = ((R * SQUARE_SIZE, D * SQUARE_SIZE), (SQUARE_SIZE, SQUARE_SIZE))
                pygame.draw.rect(DISPLAY, LIGHT_GRAY, rect, BOX_THICKNESS)

        pygame.display.flip()
        gameClock.tick(FPS)
